### PR TITLE
Add many-to-many relationship between users and organizations

### PR DESCRIPTION
**Pull request by mattwr18**
### What changed in this PR and why

This introduces the many-to-many relationship that is a requirement for some of our clients to belong to more than one organization.

Instead of a simply HABTM, it uses `has_many, through:`. This gives us more flexibility to save information on the relationship itself. For example, how do we handle active/inactive state. One idea would be to add a `deactivated_at` in the `users_organizations` table.

### TODOs

- [ ] Add automated tests that test for a user that belongs to multiple organizations
- [x] Decide on the deactivated/reactivated logic
- [x] When an admin is creating a new user, try to find the user by their email first and do something - ie either simply add the user to the organization, or throw a validation error and direct the admin to edit the existing user. EDIT: The email is already validated for uniqueness by the clearance gem. An error is displayed if someone tries to create a user with the same email.

### CODE REVIEW

**Comment by soey** [comment]
Personally I would start with the most simple version (habtm) and only extend it to a has_many_through when it is really needed. But I have no hard feeling either against it.

File: app/models/organization.rb
Comment (soey) [inline app/models/organization.rb:13]: Would that destroy a user if the organization is destroyed? I think we should keep the user as it can also belong to other organizations. The standard behavior of this should be to do nothing on the user itself and just destroy the relation.

File: app/models/user.rb
Comment (soey) [inline app/models/user.rb:13]: same as above. destroying a user should never destroy the organization.

File: db/migrate/20240812102032_create_users_organizations.rb
Comment (soey) [inline db/migrate/20240812102032_create_users_organizations.rb:3]: we might need a follow up migration to remove the organization_id from the user

File: spec/factories/users_organizations.rb
Comment (soey) [inline spec/factories/users_organizations.rb:1]: that would probably be obsolete if we would stick with a simple habtm for now..

File: spec/models/users_organization_spec.rb
Comment (soey) [inline spec/models/users_organization_spec.rb:6]: ;) habtm unless we have something here?

**Review by soey — CHANGES_REQUESTED** [review]
Please see the remarks and please (as we talked about toray) add e.g. "user_permitted?" into the ApplicationController called by the setOrganization to check if a user is allowed to access the set organization.

File: spec/models/users_organization_spec.rb
Comment (roschaefer) [reply spec/models/users_organization_spec.rb:6]: @soey I thought you and I we both agreed that it would be the cleanest solution to have a join model between users and organization? I wouldn't be against starting with `has_many, through:`

**Comment by roschaefer** [comment]
@mattwr18 I'm willing to continue on this PR and address the requested changes, unless you want to do it.

**Comment by mattwr18** [comment]
@roschaefer I don't know which changes I'll actually implement. I don't think changing to a HABTM association for no real benefit at the expense of flexibility is something I would. There are also other changes that we discussed that @soey didn't mention. We talked about the need for a validation on user email, for example.

File: spec/models/users_organization_spec.rb
Comment (soey) [reply spec/models/users_organization_spec.rb:6]: It's ok for me.

Comment (soey) [reply spec/models/users_organization_spec.rb:6]: @roschaefer: I think you are referring to the relationship between contributor and organization .. where we probably have to implement the note and mount the taggings  to

File: db/migrate/20240812102032_create_users_organizations.rb
Comment (mattwr18) [reply db/migrate/20240812102032_create_users_organizations.rb:3]: Yes, we do need one. We should create an issue for this before you merge/deploy this.

File: app/models/organization.rb
Comment (soey) [inline app/models/organization.rb:13]: can't we just leave the dependent param out. To my understanding the users is just kind of a shortcut going through the users_organization. If the users_organizations is destroyed (which is correct) then there is nothing to be done at the users at all.

Comment (mattwr18) [reply app/models/organization.rb]: We can. I thought that rubocop would complain and I think that raising an exception is a valid thing to do in this case, but the default is nil, which also makes sense. The thing is there is should be no UI to delete a user as this  leads to an exception if there are any requests by that user.

```
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "users" violates foreign key constraint "fk_rails_8ead8b1e6b" on table "requests"
DETAIL:  Key (id)=(5) is still referenced from table "requests".
```

There is also no UI to delete an organization.

See 8e2b1fb0

File: db/migrate/20240812102032_create_users_organizations.rb
Comment (soey) [reply db/migrate/20240812102032_create_users_organizations.rb:3]: #1977

File: app/controllers/application_controller.rb
Comment (soey) [inline app/controllers/application_controller.rb:38]: just as an idea for a different possible implementation:
```
def set_organization
   collection = current_user.admin? ? Organization : current_user.organizations
   collection.find(params[:organization_id]) if params[:organization_id].present?
   ...
```

the additional user_permitted? method would then be obsolete. What I like on your approach is that is quite explicit on checking but it has some overhead compared to the above.

File: db/data/20240812102718_migrate_users_to_users_organizations.rb
Comment (soey) [inline db/data/20240812102718_migrate_users_to_users_organizations.rb:5]: I was just wondering about L52 in User.rb. With that in place I assume that there are situations where an admin also has the organization_id set. We should decide if we either remove that admin scope there or include the admins in the migration as well.

File: spec/factories/users_organizations.rb
Comment (soey) [inline spec/factories/users_organizations.rb:5]: just for my interest .. why is this { nil } here?

File: spec/requests/sessions_request_spec.rb
Comment (soey) [inline spec/requests/sessions_request_spec.rb:10]: why a let and a separate org creation here?

**Review by soey — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb
Comment (soey) [inline app/controllers/application_controller.rb:5]: I would just reorder to
`  before_action :require_login, :require_otp_setup, :set_organization, :user_permitted?`

**Review by soey — CHANGES_REQUESTED** [review]
just some minor change proposals to the application controller- all other is good for now.

File: app/controllers/organizations_controller.rb
Comment (mattwr18) [reply app/controllers/organizations_controller.rb:11]: because we haven't opened a PR that moves from setting the organization with a form to simply displaying links.
It should be removed in that PR.

File: spec/factories/users_organizations.rb
Comment (mattwr18) [reply spec/factories/users_organizations.rb:5]: this was auto generated when I did

```
be rails g model UsersOrganizations user:references organization:references
```

we could maybe remove the factory as it is unused.

**Review by soey — APPROVED** [review]

**Pull request merged**
